src/main/java/dev/aziz/grocerystore/handlers/PriceHandler.java

```diff
@@ -1,27 +1,57 @@
 package dev.aziz.grocerystore.handlers;
 
 import dev.aziz.grocerystore.entities.BasketItem;
+import dev.aziz.grocerystore.entities.UserPromotion;
+import dev.aziz.grocerystore.exceptions.AppException;
+import dev.aziz.grocerystore.repositories.UserPromotionRepository;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
 
 import java.math.BigDecimal;
 import java.math.BigInteger;
+import java.time.Instant;
 import java.util.List;
 
+@Slf4j
 public class PriceHandler {
 
+    private final UserPromotionRepository userPromotionRepository;
+
     private List<BasketItem> basketItems;
 
-    public PriceHandler(List<BasketItem> basketItems) {
+    public PriceHandler(UserPromotionRepository userPromotionRepository, List<BasketItem> basketItems) {
```

Comment:
Don't inject any Spring repository here. Try to leave this class empty from Spring. Instead, add the User's promotions as input parameter